src/Introspection.hack

```diff
@@ -0,0 +1,289 @@
+        return new __Schema($schema);
+    }
+
+    <<\Slack\GraphQL\QueryRootField('__type', 'Introspect a specific type')>>
+    public static function getType(classname<IntrospectableSchema> $schema, string $name): ?__Type {
+        $type = $schema::getType($name);
+        return $type is nonnull ? new __Type($type) : null;
+    }
+}
+
+<<\Slack\GraphQL\EnumType('__TypeKind', '__TypeKind')>>
+enum __TypeKind: string {
+    SCALAR = 'SCALAR';
+    OBJECT = 'OBJECT';
+    INTERFACE = 'INTERFACE';
+    UNION = 'UNION';
+    ENUM = 'ENUM';
+    INPUT_OBJECT = 'INPUT_OBJECT';
+    LIST = 'LIST';
+    NON_NULL = 'NON_NULL';
+}
+
+<<\Slack\GraphQL\ObjectType('__Schema', 'Schema introspection')>>
+final class __Schema {
+    public function __construct(private classname<IntrospectableSchema> $schema) {}
+
+    <<\Slack\GraphQL\Field('types', 'Types contained within the schema')>>
+    public function getTypes(): vec<__Type> {
+        $schema = $this->schema;
+        return Vec\map($schema::getTypes(), $type ==> new __Type($type));
+    }
+
+    <<\Slack\GraphQL\Field('queryType', 'Query root type')>>
+    public function getQueryType(): __Type {
+        $schema = $this->schema;
+        return new __Type($schema::getQueryType());
+    }
+
+    <<\Slack\GraphQL\Field('mutationType', 'Mutation root type')>>
+    public function getMutationType(): ?__Type {
+        $schema = $this->schema;
+        $mutation_type = $schema::getMutationType();
+        return $mutation_type is nonnull ? new __Type($mutation_type) : null;
+    }
+}
+
+<<\Slack\GraphQL\ObjectType('__Type', 'Type introspection')>>
+final class __Type {
+    public function __construct(private IntrospectableType $inner_type) {}
+
+    <<\Slack\GraphQL\Field('kind', 'Kind of the type')>>
+    public function getKind(): __TypeKind {
+        $inner_type = $this->inner_type;
+        return $inner_type->getTypeKind();
+    }
+
+    <<\Slack\GraphQL\Field('name', 'Name of the type')>>
+    public function getName(): ?string {
+        $inner_type = $this->inner_type;
+        return $inner_type->getName();
+    }
+
+    <<\Slack\GraphQL\Field('description', 'Description of the type')>>
+    public function getDescription(): ?string {
+        $inner_type = $this->inner_type;
+        return $inner_type->getDescription();
+    }
+
+    <<\Slack\GraphQL\Field('fields', 'Fields in the type, only applies to OBJECT and INTERFACE')>>
+    public function getFields(bool $include_deprecated = false): ?vec<__Field> {
+        if ($this->inner_type is IntrospectableObject || $this->inner_type is IntrospectableInterface) {
+            $fields = $this->inner_type->getFields($include_deprecated);
+            return $fields is nonnull ? Vec\map($fields, $field ==> new __Field($field)) : null;
+        }
+
+        return null;
+    }
+
+    <<\Slack\GraphQL\Field('interfaces', 'Interfaces the object implements, only applies to OBJECT')>>
+    public function getInterfaces(): ?vec<__Type> {
```

Comment:
What's behind the decision for these methods to have nullable returns types? (Instead of just returning an empty vec)

Reply:
This is from the spec: https://spec.graphql.org/June2018/#sec-Schema-Introspection. My guess is it is because only certain types implement this, so for those it should return an empty vec (ie object types with no interfaces) but for any other type it should be null since the property doesn't apply to it.